csrc/scheduler/matmul_utils.cpp

```diff
@@ -294,6 +315,161 @@ std::string isMatmulFusionDefinitionSupported(
   return "";
 }
 
+// Assume that tens has a contiguous dimension, and that we will load rows of
+// that dimension, without merging with another dimension first. Then determine
+// the maximum vectorization that can be used.
+//
+// If the argument has no contiguous dimensions, then a vectorization width of 1
+// is returned.
+//
+// The sizes and strides given should be in the same order as one another and
+// should match the no-reductions allocation domain of tv.
+//
+// These rows can start at any multiple of the non-contiguous strides, so we
+// seek the largest power of 2 that divides all those other dimensions (capped
+// to 16) as well as the data pointer.
+int64_t maxRowVectorization(
+    TensorView* tv,
+    const int64_t data_ptr_int,
+    const std::vector<int64_t>& sizes,
+    const std::vector<int64_t>& strides) {
+  int64_t vec_size = scheduler_utils::maxVectorizationWidth(data_ptr_int);
+  vec_size = std::min(vec_size, 16l);
+  vec_size /= dataTypeSize(tv->dtype());
+  vec_size = std::max(vec_size, 1l);
+  if (vec_size == 1l) {
+    return vec_size;
+  }
+
+  NVF_ERROR(sizes.size() == strides.size());
+  NVF_ERROR((int64_t)sizes.size() == tv->nDims());
+
+  std::vector<std::optional<bool>> noreductions_contig;
+  for (size_t i : c10::irange(tv->getMaybeAllocationDomain().size())) {
+    if (tv->getMaybeAllocationDomain()[i]->isReduction()) {
+      continue;
+    }
+    noreductions_contig.push_back(tv->getContiguity().at(i));
+  }
+
+  bool inner_contiguous = false;
+  for (size_t i : c10::irange(sizes.size())) {
+    int64_t stride = strides.at(i);
+    int64_t size = sizes.at(i);
+    if (stride == 1) {
+      std::optional<bool> contig = noreductions_contig.at(i);
+      if (contig.has_value() && !contig.value()) {
+        // If TensorView is marked discontiguous in inner dimension, we cannot
+        // vectorize regardless of input.
+        return 1l;
+      }
+      if (inner_contiguous) {
+        // There was already a contiguous inner dimension, so that means there
+        // are multiple stride==1 dimensions in this tensor. This means that
+        // there will be overlapping rows offset from one another by a single
+        // element. In such cases we cannot vectorize. An exception is when an
+        // outer dimension has both stride==1 and size==1, in which case we
+        // could ignore its stride. Currently we ignore that case and just
+        // return 1 whenever we detect multiple stride==1 dimensions.
+        return 1l;
+      }
+      inner_contiguous = true;
+      vec_size =
+          std::min(vec_size, scheduler_utils::maxVectorizationWidth(size));
+    } else {
```

Comment:
Would it be simpler if we just do
```suggestion
    if (!*noreductions_contig.back()) {
      return 1;
    }
    if (i == sizes.size() - 1) {
      vec_size =
          std::min(vec_size, scheduler_utils::maxVectorizationWidth(size));
    } else {
```

Reply:
Good call on moving the inner contiguity check out of the loop. I simplified this code a bit in the latest push.